core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/controller/Workflow.scala

```diff
@@ -150,177 +55,166 @@ class Workflow(
+//  def getDirectUpstreamOperators(opID: OperatorIdentity): Iterable[OperatorIdentity] =
+//    inLinks.getOrElse(opID, Set())
+//
+//  def getDirectDownStreamOperators(opID: OperatorIdentity): Iterable[OperatorIdentity] =
+//    outLinks.getOrElse(opID, Set())
 
-  def getOperator(opID: OperatorIdentity): OpExecConfig = operatorToOpExecConfig(opID)
+  def getAllOperators: Iterable[NewOpExecConfig] = physicalPlan.operators
 
-  def getDirectUpstreamOperators(opID: OperatorIdentity): Iterable[OperatorIdentity] =
-    inLinks.getOrElse(opID, Set())
-
-  def getDirectDownStreamOperators(opID: OperatorIdentity): Iterable[OperatorIdentity] =
-    outLinks.getOrElse(opID, Set())
-
-  def getAllOperators: Iterable[OpExecConfig] = operatorToOpExecConfig.values
-
-  def getWorkerInfo(id: ActorVirtualIdentity): WorkerInfo = workerToLayer(id).workers(id)
+  def getWorkerInfo(id: ActorVirtualIdentity): WorkerInfo = workerToOpExecConfig(id).workers(id)
 
   /**
     * Returns the worker layer of the upstream operators that links to the `opId` operator's
     * worker layer.
     */
   def getUpStreamConnectedWorkerLayers(
-      opID: OperatorIdentity
-  ): mutable.HashMap[OperatorIdentity, WorkerLayer] = {
-    val upstreamOperatorToLayers = new mutable.HashMap[OperatorIdentity, WorkerLayer]()
-    getDirectUpstreamOperators(opID).map(uOpID =>
-      upstreamOperatorToLayers(uOpID) = getOperator(uOpID).topology.layers.last
-    )
+      opID: LayerIdentity
+  ): mutable.HashMap[LayerIdentity, NewOpExecConfig] = {
+    val upstreamOperatorToLayers = new mutable.HashMap[LayerIdentity, NewOpExecConfig]()
+    physicalPlan
+      .getUpstream(opID)
+      .foreach(uOpID => upstreamOperatorToLayers(uOpID) = physicalPlan.operatorMap(opID))
     upstreamOperatorToLayers
   }
 
-  def getSourceLayers: Iterable[WorkerLayer] = {
-    val tos = getAllLinks.map(_.to).toSet
-    getAllLayers.filter(layer => !tos.contains(layer))
-  }
-
-  def getSinkLayers: Iterable[WorkerLayer] = {
-    val froms = getAllLinks.map(_.from).toSet
-    getAllLayers.filter(layer => !froms.contains(layer))
-  }
+//  def getSourceLayers: Iterable[WorkerLayer] = {
+//    val tos = getAllLinks.map(_.to).toSet
+//    getAllLayers.filter(layer => !tos.contains(layer))
+//  }
+//
+//  def getSinkLayers: Iterable[WorkerLayer] = {
+//    val froms = getAllLinks.map(_.from).toSet
+//    getAllLayers.filter(layer => !froms.contains(layer))
+//  }
 
-  def getAllLayers: Iterable[WorkerLayer] = operatorToOpExecConfig.values.flatMap(_.topology.layers)
+//  def getAllLayers: Iterable[WorkerLayer] = operatorToOpExecConfig.values.flatMap(_.topology.layers)
 
-  def getAllLinks: Iterable[LinkStrategy] = idToLink.values
+//  def getAllLinks: Iterable[LinkStrategy] = idToLink.values
 
-  def getWorkerLayer(workerID: ActorVirtualIdentity): WorkerLayer = workerToLayer(workerID)
+  def getWorkerLayer(workerID: ActorVirtualIdentity): NewOpExecConfig =
```

Comment:
possible duplication of `getOperator(workerId: ActorVirtualIdentity)`?

Reply:
yes this is a duplication, this is removed